linkerd/detect/src/lib.rs

```diff
@@ -115,27 +104,38 @@ where
         let mut buf = BytesMut::with_capacity(self.capacity);
         let detect = self.detect.clone();
         let target = self.target.clone();
+        let timeout = self.timeout;
         Box::pin(async move {
             trace!("Starting protocol detection");
             let t0 = time::Instant::now();
-            let protocol = detect.detect(&mut io, &mut buf).await?;
-            debug!(
-                ?protocol,
-                elapsed = ?(time::Instant::now() - t0),
-                "Detected"
-            );
+
+            let protocol = match time::timeout(timeout, detect.detect(&mut io, &mut buf)).await {
+                Ok(Ok(protocol)) => {
+                    debug!(
+                        ?protocol,
+                        elapsed = ?(time::Instant::now() - t0),
+                        "Detected"
+                    );
+                    protocol
+                }
+                Ok(Err(e)) => return Err(e),
```

Comment:
nit/TIOLI: could avoid the separate `Ok(Ok(_)` and `Ok(Err(_))` cases, like this:
```suggestion
                Ok(x) => {
                    let protocol = x?;
                    debug!(
                        ?protocol,
                        elapsed = ?(time::Instant::now() - t0),
                        "Detected"
                    );
                    protocol
                }
```
but it doesn't really matter